target/riscv/cpu_helper.c

```diff
@@ -609,16 +609,23 @@ static int get_physical_address(CPURISCVState *env, hwaddr *physical,
         } else if (ppn & ((1ULL << ptshift) - 1)) {
             /* Misaligned PPN */
             return TRANSLATE_FAIL;
-        } else if (access_type == MMU_DATA_LOAD && !((pte & PTE_R) ||
-                   ((pte & PTE_X) && mxr))) {
+        } else if ((access_type == MMU_DATA_LOAD ||
+                    access_type == MMU_DATA_CAP_LOAD) &&
+                   !((pte & PTE_R) || ((pte & PTE_X) && mxr))) {
             /* Read access check failed */
             return TRANSLATE_FAIL;
-        } else if (access_type == MMU_DATA_STORE && !(pte & PTE_W)) {
+        } else if ((access_type == MMU_DATA_STORE ||
+                    access_type == MMU_DATA_CAP_STORE) && !(pte & PTE_W)) {
             /* Write access check failed */
             return TRANSLATE_FAIL;
         } else if (access_type == MMU_INST_FETCH && !(pte & PTE_X)) {
             /* Fetch access check failed */
             return TRANSLATE_FAIL;
+#if defined(TARGET_CHERI) && !defined(TARGET_RISCV32)
+        } else if (access_type == MMU_DATA_CAP_STORE && !(pte & PTE_SC)) {
```

Comment:
I'm planning to remove separate lookups for cap vs data stores and use QEMU's TLB flags instead in the future, but for now this is probably the best place for the checks.